src/librustc/ty/layout.rs

```diff
@@ -687,34 +687,117 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
             }
 
             // SIMD vector types.
-            ty::Adt(def, ..) if def.repr.simd() => {
-                let element = self.layout_of(ty.simd_type(tcx))?;
-                let count = ty.simd_size(tcx) as u64;
-                assert!(count > 0);
-                let scalar = match element.abi {
-                    Abi::Scalar(ref scalar) => scalar.clone(),
-                    _ => {
-                        tcx.sess.fatal(&format!("monomorphising SIMD type `{}` with \
-                                                 a non-machine element type `{}`",
-                                                ty, element.ty));
+            ty::Adt(def, substs) if def.repr.simd() => {
+                // Supported SIMD vectors are homogeneous ADTs with at least one field:
+                //
+                // * #[repr(simd)] struct S(T, T, T, T);
+                // * #[repr(simd)] struct S { x: T, y: T, z: T, w: T }
+                // * #[repr(simd)] struct S([T; 4])
+                //
+                // where T is a "machine type", e.g., `f32`, `i64`, `*mut _`.
+
+                // SIMD vectors with zero fields are not supported:
+                if def.non_enum_variant().fields.is_empty() {
+                    tcx.sess.fatal(&format!(
+                        "monomorphising SIMD type `{}` of zero length", ty
+                    ));
+                }
+
+                // Type of the first ADT field:
+                let f0_ty = def.non_enum_variant().fields[0].ty(tcx, substs);
+
+                // Heterogeneous SIMD vectors are not supported:
+                for fi in &def.non_enum_variant().fields {
+                    if fi.ty(tcx, substs) != f0_ty {
+                        tcx.sess.fatal(&format!(
+                            "monomorphising heterogeneous SIMD type `{}`", ty
```

Comment:
Are these new post-monomorphization errors... if so, why that?

Reply:
Not really, typeck errors before any of these can trigger, at least, as long as everything is correct there.

I added these to help during development. If typeck fails to error, an error here is better than an error down the line.

Reply:
Alright; this sounds good. Could you add comments to the places where you added these `.fatal` errors noting that these should be caught in typeck?

(Also, why isn't `bug!(...)` being used for ICEs here?)

Reply:
> (Also, why isn't bug!(...) being used for ICEs here?)

No idea, the code was using fatal error for ICES already so I just also used that - the existing fatal error even has a test..

> Could you add comments to the places where you added these .fatal errors noting that these should be caught in typeck?

Sure.

Reply:
So i've done that. 

The already existing error that fails if the element type isn't a machine type is currently not caught in typeck (this was already the case), and can fail if a generic SIMD vector is instantiated with an inappropriate element type. The way this is handled is that `libcore` should have a trait that bounds the accepted types so that this cannot happen.

There is a new monomorphization type error here now that we support generic lengths, and that's if the length passed to the vector is zero. We can only catch this once we know the actual value of the vector length, which right now is only at monomorphization time. 

Either libcore would bound the length of the vector, once we have bounds for const generics, or we should start accepting zero-element vectors (I don't see a reason to forbid this), but that is probably better done in a different PR.

Reply:
> Either libcore would bound the length of the vector, once we have bounds for const generics, or we should start accepting zero-element vectors (I don't see a reason to forbid this), but that is probably better done in a different PR.

It's also unclear at this time whether we want to support non-powers-of-two; right now we support them, this PR does not change that.

Reply:
> No idea, the code was using fatal error for ICES already so I just also used that - the existing fatal error even has a test..

`fatal` errors *are just errors* (that also abort the compilation early).
Only `(span_)bug` and panics are ICEs.

Reply:
@eddyb Is `fatal` the preferred way to omit monomorphization time errors?

Reply:
Yes, although ideally there shouldn't be any...

Reply:
@centril so is this issue resolved?

Reply:
I leave that to @eddyb -- the only thing I would note is that we don't want to add new monomorphization errors in stable but such a consideration seems far off since this is just for experimentation.

Reply:
There is a huge difference between monomorphization time errors that affect Rust users, and adding code to catch logic errors in rustc early. 

I regularly get LLVM errors when using Rust. Their output is completely useless, and the first step into debugging those is requiring users to re-compile a Rust toolchain with LLVM assertions enabled, and debugging back to Rust from there. 

I'd much rather have a monomorphization time error just saying "This generic intrinsics does not support that type", and wish others would as well.

Reply:
I think I know what needs to be cleared up: the `fatal` was moved from one place to another, it's not new (AFAIK).

Reply:
@gnzlbg I think you misunderstood me; I was making a general remark about user exposed monmorphization time errors, not internal ones for intrinsics... so we should be good.